producer/types.go

```diff
@@ -0,0 +1,152 @@
+// It should only be finalized by the producer.
+type Data interface {
+	// Shard returns the shard of the data.
+	Shard() uint32
+
+	// Bytes returns the bytes of the data.
+	Bytes() []byte
+
+	// Size returns the size of the bytes of the data.
+	Size() uint32
+
+	// Finalize will be called by producer to indicate the end of its lifecycle.
+	Finalize(DataFinalizeReason)
+}
+
+// Producer produces data to a topic.
+type Producer interface {
+	// Produce produces the data.
+	Produce(data Data) error
+
+	// RegisterFilter registers a filter to a consumer service.
+	RegisterFilter(sid services.ServiceID, fn FilterFunc)
+
+	// RemoveFilter removes the filter of a consumer service.
+	RemoveFilter(sid services.ServiceID)
+
+	// Init initializes a producer.
+	Init()
+
+	// Close stops the producer from accepting new requests immediately.
+	// It will also block until all the data buffered in Producer has been consumed.
+	Close()
+}
+
+// FilterFunc can filter data.
+type FilterFunc func(data Data) bool
+
+// Options configs a producer.
+type Options interface {
+	// Buffer returns the buffer.
+	Buffer() Buffer
+
+	// SetBuffer sets the buffer.
+	SetBuffer(value Buffer) Options
+
+	// Writer returns the writer.
+	Writer() Writer
+
+	// SetWriter sets the writer.
+	SetWriter(value Writer) Options
+}
+
+// Buffer buffers all the data in the producer.
+type Buffer interface {
+	// Buffer buffers a data and returns a reference counted data.
+	Buffer(data Data) (RefCountedData, error)
+
+	// Init initializes the buffer.
+	Init()
+
+	// Close stops the buffer from accepting new requests immediately.
+	// It will also block until all the data buffered has been consumed.
+	Close()
+}
+
+// Writer writes all the data out to the consumer services.
+type Writer interface {
+	// Write writes a reference counted data out.
+	Write(shard uint32, d RefCountedData) error
+
+	// RegisterFilter registers a filter to a consumer service.
+	RegisterFilter(sid services.ServiceID, fn FilterFunc)
+
+	//RemoveFilter removes the filter of a consumer service.
+	RemoveFilter(sid services.ServiceID)
+
+	// Init initializes a writer.
+	Init() error
+
+	// Close closes the writer.
+	Close()
+}
+
+// RefCountedData is a reference counted data.
+type RefCountedData interface {
+	// IncRef increments the ref count.
+	IncRef()
+
+	// DecRef decrements the ref count.
+	DecRef()
+
+	// IsClosed returns true if the underlying data has been closed.
+	IsClosed() bool
+
+	// Filter filters the data.
+	Filter(fn FilterFunc) bool
+
+	// Bytes returns:
+	// - underlying bytes for the data,
+	// - a bool to indicate if the bytes are valid,
+	// - a DoneFn to be called after the bytes are used.
+	// The DoneFn is used to prevent the underlying bytes
+	// being recycled while the user is still using it.
+	Bytes() ([]byte, bool, DoneFn)
+
+	// Size returns the size of the data.
+	Size() uint64
+
+	// Drop drops the data without waiting for it to be acked by consumers.
+	Drop()
```

Comment:
Perhaps the Drop function could return the bytes that are in transit. This might be handy if there is some kind of emergency where we need to drop but want to log out what was in flight when this was called. Granted this will be encoded data but having some concept of what was lost might be handy.

Reply:
Don't think the caller would need the bytes here, also it's not safe because the underlying bytes might be returned to pool and being reused already